### PR TITLE
kv/concurrency: release latches before existing lock wait-queues

### LINKED ISSUE #65429: kv/kvserver/concurrency: TestConcurrencyManagerBasic failed
kv/kvserver/concurrency.TestConcurrencyManagerBasic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2994818&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2994818&tab=artifacts#/) on master @ [64dad5f195e910196ce6c472b6cd02d4c260317c](https://github.com/cockroachdb/cockroach/commits/64dad5f195e910196ce6c472b6cd02d4c260317c):

```
=== RUN   TestConcurrencyManagerBasic
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestConcurrencyManagerBasic449174360
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestConcurrencyManagerBasic
    concurrency_manager_test.go:499: -- test log scope end --
--- FAIL: TestConcurrencyManagerBasic (4.97s)
=== RUN   TestConcurrencyManagerBasic/wait_self
    concurrency_manager_test.go:91: 
        testdata/concurrency_manager/wait_self:123: 
        output didn't match expected:
        @@ -1,7 +1,8 @@
         [-] finish reqTxnMiddle: finishing request
         [4] sequence reqTxn2: lock wait-queue event: done waiting
         [4] sequence reqTxn2: conflicted with 00000003-0000-0000-0000-000000000000 on "k" for 1.234s
         [4] sequence reqTxn2: acquiring latches
        +[4] sequence reqTxn2: waiting to acquire write latch k@20.000000000,1, held by write latch k@10.000000000,1
         [4] sequence reqTxn2: scanning lock table for conflicting locks
         [4] sequence reqTxn2: sequencing complete, returned guard
         
    --- FAIL: TestConcurrencyManagerBasic/wait_self (0.18s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestConcurrencyManagerBasic PKG=./pkg/kv/kvserver/concurrency TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/kv nvanbenschoten
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestConcurrencyManagerBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

### PR DESCRIPTION
**Pull request by nvb**
Fixes #65429.

This PR releases latches _before_ exiting lock wait-queues deliberately.
Either order would be correct, but the order here avoids non-determinism in
cases where a request holds both latches and lock wait-queue reservations
and has a request waiting on its reservations. If we were to release the
reservations before releasing latches, it would be possible for the waiting
request to beat us to the latch manager and end up blocking on our latches
briefly. Not only is this confusing in traces, but it is slightly less
efficient than if we released latches before letting anyone waiting on us
in the lock table proceed, ensuring that they do not hit our latches.

Elsewhere, we relate the relationship of between the latch manager and the
lock-table to that of a mutex and condition variable pair. Following that
analogy, this release ordering is akin to signaling a condition variable
after releasing its associated mutex. Doing so ensures that whoever the
signaler wakes up (if anyone) will never bump into its mutex immediately
upon resumption.

### CODE REVIEW

**Review by sumeerbhola — APPROVED** [review]
:lgtm: 

I didn't realize there was a flaky test -- wish there was a way of being cc'd when issues get opened on the concurrency package.

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65914#-:-Mb7hcxnCzscCdTKkTzo:b-x3dlb4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911 and @nvanbenschoten)

___
*[pkg/kv/kvserver/concurrency/concurrency_manager.go, line 301 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65914#-Mb7h6UZ7qWDfNpQI8W_:-Mb7h6UZ7qWDfNpQI8Wa:b-8ejg7b) ([raw file](https://github.com/cockroachdb/cockroach/blob/023e36b4801d78aff4e7981c6b967be507e11007/pkg/kv/kvserver/concurrency/concurrency_manager.go#L301)):*
> ```Go
> 	// NOTE: we release latches _before_ exiting lock wait-queues deliberately.
> 	// Either order would be correct, but the order here avoids non-determinism in
> 	// cases where a request holds both latches and lock wait-queue reservations
> ```

could you rephrase this in terms of request A and request B. It would be easier for a reader, compared to the use of pronouns ("us" , "we").

**Review by aliher1911 — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65914#-:-Mb7qpJu8NL7dW97wR02:b593w57)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @nvanbenschoten)

**Review by nvb — COMMENTED** [review]
TFTR!

bors r+

> wish there was a way of being cc'd when issues get opened on the concurrency package.

If you're interested, we could add you to the `@cockroachdb/kv` team, since you work so close to this area. Or we could add `@cockroachdb/storage` [as a co-owner](https://github.com/cockroachdb/cockroach/blob/2c5365dab8ee4fe6a91b64464fbd73c76c2224db/.github/CODEOWNERS#L73) of the `/pkg/kv/kvserver/concurrency` package.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/65914#-:-MbNqxHn5OfmvWbD4PYc:beservp)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @sumeerbhola)

___
*[pkg/kv/kvserver/concurrency/concurrency_manager.go, line 301 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/65914#-Mb7h6UZ7qWDfNpQI8W_:-MbNrLbpDfpFdMIm1aII:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/023e36b4801d78aff4e7981c6b967be507e11007/pkg/kv/kvserver/concurrency/concurrency_manager.go#L301)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

could you rephrase this in terms of request A and request B. It would be easier for a reader, compared to the use of pronouns ("us" , "we"). 
</blockquote></details>

Done.
